File: app/src/main/java/com/pgssoft/mvvm/views/adapters/BindingRecyclerAdapter.java
package com.pgssoft.mvvm.views.adapters;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.pgssoft.mvvm.viewmodels.BaseListItemViewModel;

import java.util.List;

/**
 * Created by bstokrocki on 30.01.2017.
 */
public abstract class BindingRecyclerAdapter<TData, TViewModel extends BaseListItemViewModel<TData>>
        extends RecyclerView.Adapter<BindingViewHolder<TViewModel>> {
    private Context context;
    private List<TData> dataItems;
    private final int variable;
    private final int layoutId;

    public BindingRecyclerAdapter(Context context, List<TData> dataItems, int variable, int layoutId) {
        this.context = context;
        this.dataItems = dataItems;
        this.variable = variable;
        this.layoutId = layoutId;
    }

    @Override
    public BindingViewHolder<TViewModel> onCreateViewHolder(ViewGroup parent, int viewType) {
        ViewDataBinding binding = DataBindingUtil.inflate(LayoutInflater.from(context),
                layoutId, parent, false);

        return new BindingViewHolder<>(binding, createViewModel(), variable);
    }

    @Override
    public void onBindViewHolder(BindingViewHolder<TViewModel> holder, int position) {
        TData dataItem = dataItems.get(position);
        holder.getViewModel().setData(dataItem);
    }

    @Override
    public int getItemCount() {
        return dataItems.size();
    }

    public abstract TViewModel createViewModel();
}

File: kotlin/app/src/main/java/com/pgssoft/kotlinplayground/view/adapter/BindingRecyclerAdapter.kt
package com.pgssoft.mvvm.views.adapters

import android.content.Context
import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.pgssoft.kotlinplayground.view.adapter.BindingViewHolder
import com.pgssoft.kotlinplayground.viewmodel.BaseListItemViewModel

/**
 * Created by bstokrocki on 30.01.2017.
 */
abstract class BindingRecyclerAdapter<TData, TViewModel : BaseListItemViewModel<TData>>(
        private val context: Context,
        protected var dataItems: List<TData>,
        private val variable: Int,
        private val layoutId: Int) : RecyclerView.Adapter<BindingViewHolder<TData, TViewModel>>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BindingViewHolder<TData, TViewModel> {
        val binding = DataBindingUtil.inflate<ViewDataBinding>(LayoutInflater.from(context),
                layoutId, parent, false)

        return BindingViewHolder(createViewModel(), binding, variable)
    }

    override fun onBindViewHolder(holder: BindingViewHolder<TData, TViewModel>, position: Int) {
        val dataItem = dataItems[position]
        holder.viewModel.setData(dataItem)
    }

    override fun getItemCount(): Int {
        return dataItems.size
    }

    abstract fun createViewModel(): TViewModel
}
